ocaml/xenopsd/xc/xenctrlext.ml

```diff
@@ -82,3 +82,35 @@ type cputopo = {core: int; socket: int; node: int}
 external numainfo : handle -> numainfo = "stub_xenctrlext_numainfo"
 
 external cputopoinfo : handle -> cputopo array = "stub_xenctrlext_cputopoinfo"
+
+module Xenforeignmemory = struct
+  type handle
+
+  (** pages are mapped linearly *)
+  type mapping =
+    (char, Bigarray.int8_unsigned_elt, Bigarray.c_layout) Bigarray.Array1.t
+
+  type prot = {read: bool; write: bool; exec: bool}
+
+  external acquire : Xentoollog.handle option -> handle
+    = "stub_xenforeignmemory_open"
+
+  external release : handle -> unit = "stub_xenforeignmemory_close"
+
+  external map : handle -> domid -> prot -> Int64.t list -> mapping
+    = "stub_xenforeignmemory_map"
+
+  external unmap : handle -> mapping -> unit = "stub_xenforeignmemory_unmap"
+
+  let with_mapping (handle : handle) (domid : domid) (prot : prot)
+      (pages : Int64.t list) (f : mapping -> unit) =
+    let mapping = map handle domid prot pages in
+    Fun.protect
+      ~finally:(fun () ->
+        try unmap handle mapping
+        with Unix_error (errno, _) ->
+          Printf.printf "Error while unmapping memory: %s\n"
```

Comment:
We shouldn't print the error message here, stdout is likely to go nowhere (/dev/null).
If we had access to a logging framework (e.g. Logs) we could call that here, but if the intention is to upstream these bindings we can't do that. Although if the intention is to upstream the bindings we can't print to stdout here either, and we can't raise in the `finally` either, and we don't want to ignore the exception either...

Perhaps add an optional parameter to `with_mapping`, e.g. `?on_unmap_failure` that can be used to log the exception.

Reply:
Yeah, I've added the parameter to log it the user wants to, I've made the parameter optional.